ops/main.py

```diff
@@ -356,9 +325,255 @@ def _should_use_controller_storage(db_path: Path, meta: CharmMeta) -> bool:
+        if charm_dir is None:
+            # Assume $JUJU_CHARM_DIR/lib/op/main.py structure.
+            charm_dir = Path(f'{__file__}/../../..').resolve()
+        else:
+            charm_dir = Path(charm_dir).resolve()
+        return charm_dir
+
+    @staticmethod
+    def autoload(charm_type: Type["ops.charm.CharmBase"]):
+        charm_root = _CharmSpec._get_charm_dir()
+
+        metadata_path = charm_root / "metadata.yaml"
+        if not metadata_path.exists():
+            raise FileNotFoundError(
+                f"invalid charm root {charm_root!r}; "
+                f"expected to contain at least a `metadata.yaml` file.",
+            )
+        meta = yaml.safe_load(metadata_path.open())
+
+        actions = config = None
+
+        config_path = charm_root / "config.yaml"
+        if config_path.exists():
+            config = yaml.safe_load(config_path.open())
+
+        actions_path = charm_root / "actions.yaml"
+        if actions_path.exists():
+            actions = yaml.safe_load(actions_path.open())
+
+        return _CharmSpec(
+            charm_type=charm_type,
+            charm_root=charm_root,
+            meta=meta,
+            actions=actions,
+            config=config
+        )
+
+
+class Ops:
+    """Initializes the Framework and manages the lifecycle of a charm.
+
+    Running Ops consists of three main steps:
+    - setup: initialize from envvars and argv's the objects that encapsulate:
+      - the Framework (hook tool wrappers)
+      - the storage backend
+      - the event that Juju is emitting on us
+      - the charm instance (user-facing)
+    - emit: core user-facing lifecycle step. Consists of:
+      - reemit any deferred events found in the storage
+      - emit the Juju event on the charm
+      - emit any other events the charm emitted on itself during this cycle
+    - commit: responsible for:
+      - store any events deferred throughout this execution
+      - graceful teardown of the storage
+      - emission of the ``collect-status`` events
+    """
+    CHARM_STATE_FILE = '.unit-state.db'
+
+    def __init__(
+            self,
+            charm_spec: "_CharmSpec",
+            use_juju_for_storage: Optional[bool] = None
+    ):
+        self.charm_spec = charm_spec
+        self._charm_meta = CharmMeta(charm_spec.meta, charm_spec.actions)
+        self._use_juju_for_storage = use_juju_for_storage
+
+        # set by setup()
+        self.dispatcher: Optional[_Dispatcher] = None
+        self.framework: Optional["ops.framework.Framework"] = None
+        self.charm: Optional["ops.charm.CharmBase"] = None
+
+        self._has_setup = False
+        self._has_emitted = False
+        self._has_committed = False
+
+    def _setup_charm(self, framework: "ops.framework.Framework", dispatcher: _Dispatcher):
+        charm_type = self.charm_spec.charm_type
+        sig = inspect.signature(charm_type)
+        sig.bind(framework)  # signature check
```

Comment:
What's the advantage of doing this here? If the signature is wrong, doesn't it just move the error to this line with `TypeError: too many positional arguments` (or whatever the bad sig is), rather than a couple of lines down with `TypeError: MyCharm.__init__() takes 1 positional argument but 2 were given` (for example).

It seems like letting it fail a couple of lines later with a clearer error that it's the `__init__` signature that's at fault is better. I assume I'm missing a benefit :)